RavenDB-23942 : Add Gen AI Integration

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23942/Add-AI-Gen-Integration

## GenAI Feature – Initial Implementation
This is the initial PR introducing the new Generative AI feature to RavenDB.

**Purpose**:
The GenAI feature enables RavenDB to integrate with LLMs to analyze and transform documents using AI-powered scripts.

**What's included so far:**

- Basic ETL Pipeline for GenAI, including transformation, model inference, and update stages.

- Integration with AI connectors, including prompt configuration, schema definition, and output handling.

- Test Mode Support: Modular test execution per stage (context creation, model call, patching).

- Metadata Hash Tracking: Automatically tracks processed content via @genAiHashes metadata, to avoid redundant AI calls.

- Initial test suite, including end-to-end test cases and metadata verification.

This PR is still a work in progress. Feedback is welcome at this stage to guide finalization and further refinements.

ayende committed:
RavenDB-23942 Starting to add AI Gen ETL (currently mostly infrastructure)

kalczur committed:
RavenDB-23955 AiGenConfiguration -> GenAiConfiguration and add it to typing gen

kalczur committed:
RavenDB-23955 Prepare empty view for edit

kalczur committed:
RavenDB-23955 AiGenConfiguration -> GenAiConfiguration

kalczur committed:
RavenDB-23955 AiGen -> GenAi

kalczur committed:
RavenDB-23955 Add some fields to GenAiConfiguration

kalczur committed:
RavenDB-23955 Add EditGenAiTask view

Commit:
RavenDB-23942 : provide a way to run GeAi ETL in test mode

Commit:
RavenDB-23942 : fix errors after rebase, add GenAiTransformation to configuration instead of using Transformations

Commit:
RavenDB-23942 : rename AiGen -> GenAi

Commit:
RavenDB-23942 : GenAI test mode - fix object disposed error, assert valid results in test

Commit:
RavenDB-23942 : GenAi test mode : add endpoints + test + adjust EtlProcess.TestScript

Commit:
RavenDB-23942 : do not open write tx for the entire test duration. instead, use read tx for the extract & transform phases, close it before sending to model, and open write tx just before the update phase.
for the regular flow (none test-mode) we'll do the same - use read tx for extract & transform and close it before sending to model (no need to reopen another tx afterwards).

Commit:
RavenDB-23942 : test caching behavior

Commit:
RavenDB-23942 : add GenAi tasks to AbstractOngoingTasks.GetAllTasks

kalczur committed:
RavenDB-23955 Fix compilation error

kalczur committed:
RavenDB-23955 Show GenAI tasks on the list

Commit:
RavenDB-23942 : handle getting GenAi ongoing task info

kalczur committed:
RavenDB-23955 Allow to edit GenAI

kalczur committed:
RavenDB-23955 Add GenAI test

kalczur committed:
RavenDB-23955 Remove identifier

kalczur committed:
RavenDB-23955 Split test results

kalczur committed:
RavenDB-23942 EditGenAiTask adjustments

Commit:
RavenDB-24089 : Make test mode modular: enable testing of each phase independently (context creation, model output, update script)

Commit:
RavenDB-23942 : handle editing/deleting GenAi task

Commit:
RavenDB-24089 : minor cleanup and test case modification

kalczur committed:
RavenDB-23942 Ignore error

kalczur committed:
RavenDB-23942 Fix NRE

kalczur committed:
RavenDB-23942 Fix raven types issue

kalczur committed:
RavenDB-23942 Task view adjustments

Commit:
RavenDB-23942 - Fully wiring end to end Ai Gen - we can define a prompt, call to the model, get the results back and apply them.

Commit:
RavenDB-23942 : minor fixes after rebase

Commit:
RavenDB-23942 ; bug fixes in ChatCompletionClient and GenAiTask

kalczur committed:
RavenDB-23942 Add steps

kalczur committed:
RavenDB-23942 UI improvements

kalczur committed:
RavenDB-23942 UI improvements

kalczur committed:
RavenDB-23942 UI improvements

Commit:
RavenDB-23942 : test-mode modifications : use TestStage enum instead of booleans, allow to pass Document to test method (so we can edit it and work on modified version)

Commit:
RavenDB-24095 : handle disable/enable GenAi task

Commit:
RavenDB-23942 : add tests for edit/delete/toggle-state

kalczur committed:
RavenDB-23942 Prepare for test

kalczur committed:
RavenDB-23942 Tabs styles

kalczur committed:
RavenDB-23942 Add playground collapse

kalczur committed:
RavenDB-23942 Add edit mode button

Commit:
RavenDB-24089 : add test cases for calling test-mode endpoint

kalczur committed:
RavenDB-23942 Improve test results

Commit:
RavenDB-24089 : test-mode modifications - allow to pass a fake document (doesn't exists in db) and don't require passing a docId in this case

kalczur committed:
RavenDB-23942 UI improvements

kalczur committed:
RavenDB-23942 Move testing state to store

Commit:
RavenDB-24089 : bug fix - do not try to create a chatCompletionClient in initial test-mode stage (there's no schema/sample object at this point). create the client only in SendToModel stage

Commit:
RavenDB-24089 : test mode bug fix - when document is provided as input we need to give it a "dummy" id and store it (without commit), otherwise the patch command won't be able to load it

kalczur committed:
RavenDB-23942 Storybook and UI improvements

kalczur committed:
RavenDB-23942 Layout updates

kalczur committed:
RavenDB-23942 Add virtualization

kalczur committed:
RavenDB-23942 Change edit mode from button to switch

kalczur committed:
RavenDB-23942 Add tooltips

kalczur committed:
RavenDB-23942 Script results styling

kalczur committed:
RavenDB-23942 Styling

kalczur committed:
RavenDB-23942 Virtual list styling

kalczur committed:
RavenDB-23942 Add tokens usage

kalczur committed:
RavenDB-23942 Add mocks

kalczur committed:
RavenDB-23942 Improve edit mode

kalczur committed:
RavenDB-23942 Fix file name

kalczur committed:
RavenDB-23942 Fix file name

Commit:
RavenDB-24093 : Move hashes to document metadata

ayende commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+
+    public string Prompt { get; set; }
+    
+    //TODO: Make this JSON objects? 
```
Yes, this should be objects, not strings

ayende commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+    //TODO: Make this JSON objects? 
+    public string JsonSchema { get; set; }
+    public string SampleObject { get; set; }
+    public string Update { get; set; }
```
UpdateScript, instead?

aviv86 replied:
done

ayende commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+    public string SampleObject { get; set; }
+    public string Update { get; set; }
+    
+    private List<Transformation> _transforms;
```
_Can_ we have multiple transforms here?

ayende commented on src/Raven.Client/Documents/Operations/AI/GenAiTransformation.cs:
```diff
@@ -0,0 +1,23 @@
+        error = string.Empty;
+        if (Script.Contains("context")) 
+            return true;
+        error = "You must call the context(ctx, hash) function in your script";
```
This is wrong now, no?

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/AI/AiGen/ChatCompletionClient.cs:
```diff
@@ -0,0 +1,169 @@
+        _client?.Dispose();
+    }
+
+    public static string GetSchemaFor(string schemaOrSampleObject)
```
This should probably be extracted

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/Embeddings/EmbeddingsGenerationTask.cs:
```diff
@@ -21,15 +16,13 @@
 namespace Raven.Server.Documents.ETL.Providers.AI.Embeddings;
 
-public sealed class EmbeddingsGenerationTask : EtlProcess<EmbeddingsGenerationItem, EmbeddingGenerationScriptResult, EmbeddingsGenerationConfiguration, AiConnectionString,
+public sealed class EmbeddingsGenerationTask : EtlProcess<AiEtlItem, EmbeddingGenerationScriptResult, EmbeddingsGenerationConfiguration, AiConnectionString,
```
Why was this changed? EmbeddingsGenerationItem -> AiEtlItem?

aviv86 replied:
changed this back to `EmbeddingsGenerationItem` and created a new type for GenAI (`GenAiItem`)

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+                _patch = (_patchRequest, CreatePatchArgs(context, item));
+
+                var patchResult = ExecuteOnDocument(context, item.DocId, expectedChangeVector: null, runner, runIfMissing: null);
+                tuple.Doc = patchResult.ModifiedDocument;
```
Does this work? `()` is a value type, are you are modifying a _copy_ of that.

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+
+                var patchResult = ExecuteOnDocument(context, item.DocId, expectedChangeVector: null, runner, runIfMissing: null);
+                tuple.Doc = patchResult.ModifiedDocument;
+                hashes[item.DocId] = tuple;
```
Better to use `CollectionMarshal` here - to avoid the two dic lookups.

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+        foreach (var kvp in hashes)
+        {
+            var doc = kvp.Value.Doc;
+            if (doc == null)
```
Under what scenario can that happen? Also - aren't we _writing_ the document here?
What happens if the `UpdateScript` _delete_ the doc?

ayende replied:
What if it didn't _modify_ the root doc, but created a new one?

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+        {
+            // no metadata at all (shouldn't happen)
+
+            doc.Modifications = new DynamicJsonValue(doc)
```
Is that even possible? Wouldn't we have at least the `@empty` one here? We already created the doc via patch, no?

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs:
```diff
@@ -0,0 +1,120 @@
+        return JsValue.Null;
+    }
+
+    private static bool ShouldSendContext(string hash, string taskName, Document doc)
```
Do we want to put it here? Or move it to the next stage, like we do in tests?

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+    {
+        var connectorType = cfg.Connection.GetActiveProvider();
+        var (uri, model, apiKey) = connectorType switch
+        {
```
I think we need to implment the rest here as well, in the same manner

ayende replied:
Always using the same client, of course.

karmeli87 replied:
I think this can be done after the release

ayende commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+                if (item.ModelOutput is null)
+                    continue; 
+
+                _patch = (_patchRequest, CreatePatchArgs(context, item));
```
I wonder if we should add an Init phase here (executed outside the tx merger) to reduce overall costs

ppekrol replied:
I agree, would be better to mirror PatchDocumentCommand(Base)

ayende commented on src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs:
```diff
@@ -33,7 +33,7 @@ public abstract class PatchDocumentCommandBase : DocumentMergedTransactionComman
         protected readonly (PatchRequest Run, BlittableJsonReaderObject Args) _patchIfMissing;
         private readonly BlittableJsonReaderObject _createIfMissing;
-        protected readonly (PatchRequest Run, BlittableJsonReaderObject Args) _patch;
+        protected (PatchRequest Run, BlittableJsonReaderObject Args) _patch;
```
Why is that not readonly?

ppekrol replied:
ping

ayende commented on src/Raven.Server/ServerWide/Commands/AI/UpdateGenAiCommand.cs:
```diff
@@ -0,0 +1,92 @@
+
+    private void InClusterValidation(DatabaseRecord record)
+    {
+        //TODO
```
Reminder that this needs to be done

ayende commented on src/Raven.Server/ServerWide/Commands/AI/AddGenAiCommand.cs:
```diff
@@ -0,0 +1,30 @@
+
+    private void Validate(DatabaseRecord databaseRecord)
+    {
+       //TODO
```
Reminder

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -2208,12 +2209,31 @@ bool CanDeleteSecretKey()
+                        var aiIntegration = JsonDeserializationCluster.GenAiConfiguration(etlConfiguration);
+                        if (string.IsNullOrEmpty(aiIntegration.JsonSchema))
+                        {
+                            // todo: move this to a better location
```
Reminder - what is the better location?

shaharhikri replied:
Done

ayende commented on test/FastTests/GenAi/GenAiBasics.cs:
```diff
@@ -0,0 +1,532 @@
+    public void CanCreateGenAiTask()
+    {
+        using var store = GetDocumentStore();
+        store.Maintenance.Send(new PutConnectionStringOperation<AiConnectionString>(new AiConnectionString
```
This will fail when Ollama isn't running - we need to first - skip it if it isn't running.
And we need to ensure that there _is_ an Ollama running, at least for the CI nightlies.

aviv86 replied:
done

ayende commented on test/FastTests/GenAi/GenAiBasics.cs:
```diff
@@ -0,0 +1,532 @@
+                Script = @"
+for(const comment of this.Comments)
+{
+    context({Text: comment.Text, Author: comment.Author, Id: comment.Id}, comment.AiHash);
```
This needs to be changed, no?

aviv86 replied:
done

ayende reviewed (commented):
Looks goot - see comments

kalczur committed:
RavenDB-23942 Add cancel to info alerts

ppekrol commented on src/Raven.Client/Documents/Operations/AI/AiConnectionString.cs:
```diff
@@ -95,6 +95,31 @@ public AiSettingsCompareDifferences Compare(AiConnectionString newConnectionStri
         return result;
     }
 
+    public bool UsingEncryptedCommunicationChannel()
```
internal?

aviv86 replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+    public override EtlType EtlType => EtlType.GenAi;
+    public override bool UsingEncryptedCommunicationChannel() => Connection?.UsingEncryptedCommunicationChannel() ?? false;
+
+    public string GenerateIdentifier() => EmbeddingsGenerationConfiguration.GenerateIdentifier(Name);
```
shouldn't be here? check how this is handled in EmbeddingsGenerationConfiguration

ppekrol replied:
internal at least?

ppekrol commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+
+    [JsonDeserializationIgnore]
+    [JsonIgnore]
+    [Obsolete($"{nameof(GenAiConfiguration)} doesn't support multiple transformations.")]
```
write which property to use instead

aviv86 replied:
done

ppekrol commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -127,29 +127,31 @@ public DatabaseRecord(string databaseName)
         
         public List<EmbeddingsGenerationConfiguration> EmbeddingsGenerations = [];
+        
+        public List<GenAiConfiguration> GenAiEtls = [];
```
is this GenAiEtl or just GenAis / GenAIs?

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/ChatCompletionClient.cs:
```diff
@@ -0,0 +1,169 @@
+using System.Text.Json.Nodes;
+using System.Threading.Tasks;
+
+namespace Raven.Server.Documents.AI.AiGen;
```
AiGen or GenAi?

ayende replied:
The name should be `GenAI` all `AiGen` should be replaced to `GenAI`

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/ChatCompletionClient.cs:
```diff
@@ -0,0 +1,169 @@
+
+namespace Raven.Server.Documents.AI.AiGen;
+
+public class ChatCompletionClient(Uri baseUri, string model, string apiKey, string structuredOutputSchema) : IDisposable
```
validate parameters

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -317,8 +320,18 @@ private IEnumerable<EtlProcess> GetRelevantProcesses<T, TConnectionString>(List<
+                        aiGenGenerationConfig = config as GenAiConfiguration;
+                        
+                        if (_databaseRecord.AiConnectionStrings.TryGetValue(config.ConnectionStringName, out var genAiConStr))
+                            aiGenGenerationConfig.Initialize(genAiConStr);
```
just `aiGenConfig` ?

karmeli87 replied:
also here should be genAi

aviv86 replied:
fixed

ppekrol commented on src/Raven.Server/Documents/ETL/EtlProcess.cs:
```diff
@@ -1195,6 +1216,29 @@ public static TestEtlScriptResult TestScript<TC, TCS>(
+            }
+
+            using var tx = testScript.IsDelete
+                ? context.OpenWriteTransaction()
```
I'm a bit worried about this, because this will hold the write tx lock

ayende replied:
Note that this is how we do that across all the ETLs that modify docs. 
This is only for testing purposes. not expected to be done on high load systems.

arekpalinski replied:
That's correct. It just got moved here.

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/AiEtlItem.cs:
```diff
@@ -2,14 +2,14 @@
 
 namespace Raven.Server.Documents.ETL.Providers.AI;
 
-public sealed class EmbeddingsGenerationItem : ExtractedItem
```
cc @arekpalinski @maciejaszyk

ppekrol replied:
wouldn't it be better to have a separate item?

maciejaszyk replied:
Well, currently there is no need for that, however, maybe in future more properties will be added for GenAI?

arekpalinski replied:
Usually we introduce a dedicated class per each ETL. It's true that they are mostly very similar / same classes. Technically we could use a generic `AiEtlItem` and split if needed in the future, although I like to have them separated from the beginning. It's not a big deal this additional class and we don't mix objects between ETLs then.

aviv86 replied:
done, we have have a separate item now

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+              collectResultsNeeded: true,
+              returnDocument: false)
+    {
+        _items = items;
```
validate the params?

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+
+        reply?.Add(new DynamicJsonValue
+        {
+            [nameof(BatchRequestParser.CommandData.Type)] = "GenAiBatchPATCH"
```
shouldn't this come from some enum or a const (I see the TODO, but FYI)?

ppekrol commented on src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForAddEtl.cs:
```diff
@@ -92,18 +92,30 @@ protected virtual void AssertCanAddOrUpdateEtl(ref BlittableJsonReaderObject etl
+                    {
+                        var aiGenConfiguration = Client.Json.Serialization.JsonDeserializationClient.AiGenConfiguration(etlConfiguration);
+                        var connectionStringName = aiGenConfiguration.ConnectionStringName ?? string.Empty;
+                        RequestHandler.ServerStore.LicenseManager.AssertCanAddEmbeddingsGenerationTask(GetAiConnectionString(context, connectionStringName));
```
AssertCanAddEmbeddingsGenerationTask?

ppekrol commented on src/Raven.Server/ServerWide/ClusterCommandsVersionManager.cs:
```diff
@@ -202,6 +202,10 @@ public sealed class ClusterCommandsVersionManager
             [nameof(PutAiConnectionStringCommand)] = 70_001,
             [nameof(RemoveAiConnectionStringCommand)] = 70_001,
+            
+            [nameof(AddGenAiCommand)] = 70_002,
```
71_000

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+        PatchRequest req = new(Configuration.Update, PatchRequestType.AiGen);
+        var cmd = new GenAiBatchPatchCommand(context, results, req, Configuration.Name);
+
+        Database.TxMerger.Enqueue(cmd).GetAwaiter().GetResult();
```
there is EnqueueSync method there

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+        return new GenAiStatsScope(stats);
+    }
+
+    protected override string StatsAggregatorTag => "Generative AI";
```
Let's stick to GenAI?

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+            var stream = new ReadOnlyMemoryStream<byte>(Encoding.UTF8.GetBytes(result));
+            item.ModelOutput = new ModelOutput
+            {
+                Output = context.ReadForMemoryAsync(stream, item.DocId).GetAwaiter().GetResult()
```
context.Sync.ReadForMemory ?

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+            // TODO: report usage
+
+            //TODO: REALLY YUCKY!
+            var stream = new ReadOnlyMemoryStream<byte>(Encoding.UTF8.GetBytes(result));
```
using?

aviv86 replied:
removed

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+            if (Configuration.TestMode == false)
+                continue;
+
+            stream = new ReadOnlyMemoryStream<byte>(Encoding.UTF8.GetBytes(usage));
```
using?

aviv86 replied:
removed

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+                continue;
+
+            stream = new ReadOnlyMemoryStream<byte>(Encoding.UTF8.GetBytes(usage));
+            item.ModelOutput.Usage = context.ReadForMemoryAsync(stream, item.DocId).GetAwaiter().GetResult();
```
context.Sync.

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,384 @@
+        else
+        {
+            if (string.IsNullOrEmpty(testGenAiScript.DocumentId))
+                throw new InvalidOperationException("Document or DocumentId must be provided to run GenAI test");
```
this is not needed for all the stages, right?

aviv86 replied:
right, will fix that in a dedicated issue

Commit:
RavenDB-24089 : test-mode : clone the test document if it has a blittable parent (happens when ran from Studio) to avoid "Attempted to copy a nested object" error

grisha-kotler committed:
RavenDB-23942 - add the GenAi license attribute

grisha-kotler committed:
RavenDB-23942 - temporarily set to true

grisha-kotler committed:
RavenDB-23942 - change to GenAi

grisha-kotler committed:
RavenDB-23942 - allow to save the connection string without having the embedding feature

grisha-kotler committed:
RavenDB-23942 - move to slow tests

grisha-kotler committed:
RavenDB-23942 - change namespace

grisha-kotler committed:
RavenDB-23942 - add a new license type

grisha-kotler committed:
RavenDB-23942 - update snmp

grisha-kotler committed:
RavenDB-23942 - fix failing test

karmeli87 commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1651,13 +1662,25 @@ public void AssertCanAddEmbeddingsGenerationTask(AiConnectionString aiConnection
                 return;
             
-            if (aiConnectionString != null && aiConnectionString.GetActiveProvider() == AiConnectorType.Embedded)
+            if (aiConnectionString == null || aiConnectionString.GetActiveProvider() == AiConnectorType.Embedded)
```
@grisha-kotler this should be also fixed in 7.0?

ppekrol replied:
ping

ppekrol replied:
@grisha-kotler ?

grisha-kotler replied:
Yes. I'll create a separate PR for this for 7.0.

kalczur committed:
RavenDB-23942 Add edit mode warning

kalczur committed:
RavenDB-23942 Add empty results

kalczur committed:
RavenDB-23942 Styling

maciejaszyk commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+        errors = [];
+
+        if (validateName && string.IsNullOrEmpty(Name))
+            errors.Add($"{nameof(Name)} of AiGen configuration cannot be empty");
```
`GenAi`?

aviv86 replied:
fixed

arekpalinski commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+        errors = [];
+
+        if (validateName && string.IsNullOrEmpty(Name))
+            errors.Add($"{nameof(Name)} of AiGen configuration cannot be empty");
```
Minor - we have `GenAiConfiguration` but the error messages say about "AiGen" - GenAi vs AiGen

aviv86 replied:
fixed

karmeli87 commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs:
```diff
@@ -0,0 +1,120 @@
+
+        foreach (var h in existingHashes)
+        {
+            if (string.Equals(hash, h?.ToString(), StringComparison.OrdinalIgnoreCase))
```
the hash is key insensitive?

karmeli87 commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -124,6 +122,7 @@ private void LoadProcesses(DatabaseRecord record,
             List<QueueEtlConfiguration> newQueueDestinations,
             List<SnowflakeEtlConfiguration> newSnowflakeDestinations,
             List<EmbeddingsGenerationConfiguration> newEmbeddingsGenerationDestinations,
+            List<GenAiConfiguration> newAiGenDestinations,
```
also here should be genAi

aviv86 replied:
fixed

karmeli87 commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs:
```diff
@@ -0,0 +1,120 @@
+
+    public override void Initialize(bool debugMode)
+    {
+        Database.Scripts.GetScriptRunner(_mainScript, true, out DocumentScript);
```
do we need to dispose this?

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs:
```diff
@@ -0,0 +1,120 @@
+    protected override void LoadToFunction(string tableName, ScriptRunnerResult colsAsObject)
+    {
+        throw new NotImplementedException();
+    }
```
better to throw not supported?

aviv86 replied:
done

maciejaszyk commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand .cs:
```diff
@@ -0,0 +1,173 @@
+
+namespace Raven.Server.Documents.ETL.Providers.AI.GenAi;
+
+internal sealed class GenAiBatchPatchCommand : PatchDocumentCommandBase
```
There is a space in the file name

aviv86 replied:
fixed, thanks

arekpalinski commented on src/Raven.Server/Documents/AI/AiGen/ChatCompletionClient.cs:
```diff
@@ -0,0 +1,169 @@
+            ["response_format"] = _schema.DeepClone()
+        };
+
+        using var reply = await _client.PostAsync("/v1/chat/completions", JsonContent.Create(req)).ConfigureAwait(false);
```
When I tried to find more info about what are "completions" I stumbled upon this:

> Starting a new project? We recommend trying [Responses](https://platform.openai.com/docs/api-reference/responses) to take advantage of the latest OpenAI platform features. Compare [Chat Completions with Responses](https://platform.openai.com/docs/guides/responses-vs-chat-completions?api-mode=responses).

https://platform.openai.com/docs/api-reference/chat

Is it intentional that we Chat Completions API here?

Lwiel replied:
OpenAI has two different APIs we could use, I also stumbled upon this recently. They expect different schemas. And what about other providers?

maciejaszyk commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+
+    protected override IEnumerator<AiEtlItem> ConvertDocsEnumerator(DocumentsOperationContext context, IEnumerator<Document> docs, string collection)
+    {
+        return new DocumentsToEmbeddingsGenerationItems(docs, collection);
```
This and other overrides contains `...EmbeddingsGenerationTask` . Please check it.

aviv86 replied:
fixed

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+    GenAiStatsScope, GenAiPerformanceOperation>
+{
+
+    internal const string GenAiHashesMetadataKey = "@genAiHashes";
```
It could be `@gen-ai-hashes` for consistency with other metadata that we already have: `@change-vector`, `@last-modified`, `@counters-snapshot`,  `@timeseries-named-values` etc

Please also consider moving it to `Raven.Client.Constants.Metadata`

aviv86 replied:
changed to `@gen-ai-hashes` and moved to `Raven.Client.Constants.Metadata`

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+        return new GenAiScriptTransformer(Database, context, Transformation, null, Configuration);
+    }
+
+    protected override string LoadFailureMessage => $"Failed to generate embeddings in '{Configuration.Name}' task. Going to do the retry in {FallbackTime} (failure #{_fallbackCounter}).";
```
The message says about "embeddings". Please adjust

aviv86 replied:
fixed

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+            _ => secondsToWaitToNextMinute + ((int)Math.Log2(_fallbackCounter) * 60)
+        };
+
+        double max = Database.Configuration.Ai.EmbeddingsGenerationMaxFallbackTime.AsTimeSpan.TotalSeconds;
```
Should it be a dedicated configuration option? Specific to Gen AI feature.

aviv86 replied:
removed this
there's no dedicated configuration option for now

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+            var stream = new ReadOnlyMemoryStream<byte>(Encoding.UTF8.GetBytes(result));
+            item.ModelOutput = new ModelOutput
+            {
+                Output = context.ReadForMemoryAsync(stream, item.DocId).GetAwaiter().GetResult()
```
Please check `ValueTask instances should not have their result directly accessed unless the instance has already completed` warning here

aviv86 replied:
fixed

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+                continue;
+
+            stream = new ReadOnlyMemoryStream<byte>(Encoding.UTF8.GetBytes(usage));
+            item.ModelOutput.Usage = context.ReadForMemoryAsync(stream, item.DocId).GetAwaiter().GetResult();
```
Same warning as above here

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+                var transformedResults = Transform([aiEtlItem], context, new GenAiStatsScope(new EtlRunStats()), new EtlProcessState());
+                items = PrepareItemsBeforeSendingToModel(transformedResults);
+
+                context.CloseTransaction();
```
Shouldn't be in `finally`?

arekpalinski commented on src/Raven.Server/Documents/Patch/PatchRequest.cs:
```diff
@@ -20,7 +20,8 @@ public enum PatchRequestType
         SnowflakeEtl,
-        EmbeddingsGeneration
+        EmbeddingsGeneration,
+        AiGen
```
GenAi?

aviv86 replied:
fixed

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/Handlers/GenAiHandler.cs:
```diff
@@ -0,0 +1,15 @@
+
+public sealed class GenAiHandler : DatabaseRequestHandler
+{
+    [RavenAction("/databases/*/admin/ai/genai/test", "POST", AuthorizationStatus.DatabaseAdmin)]
```
Maybe 
```suggestion
    [RavenAction("/databases/*/admin/ai/gen-ai/test", "POST", AuthorizationStatus.DatabaseAdmin)]
```
?

aviv86 replied:
done

arekpalinski commented on test/SlowTests/Server/Documents/AI/GenAi/GenAiBasics.cs:
```diff
@@ -0,0 +1,533 @@
+            Identifier = "ollama-local",
+            OllamaSettings = new OllamaSettings
+            {
+                Uri = "http://127.0.0.1:11434/",
```
We have `RAVEN_AI_INTEGRATION_OLLAMA_URI` defined on Jenkins. See `OllamaConnectorForTesting`

aviv86 replied:
done

maciejaszyk commented:
Studio allows me to set local model, which contains only embedder:
![image](https://github.com/user-attachments/assets/0a6773a7-e889-42c1-a6fb-48c4097b2650)

After selecting a collection I think we can show docs ids from that collection:
![image](https://github.com/user-attachments/assets/ac3ab977-dc1c-40e6-900a-2df305d15705)

arekpalinski commented on test/SlowTests/Server/Documents/AI/GenAi/GenAiBasics.cs:
```diff
@@ -0,0 +1,533 @@
+
+namespace SlowTests.Server.Documents.AI.GenAi;
+
+public class GenAiBasics(ITestOutputHelper output) : RavenTestBase(output)
```
Please add a test for export / import (backup / restore) of a database with Gen AI task defined.

I didn't see any changes in the Smuggler code so I believe this is still TODO.

aviv86 replied:
done
see `GenAiBackupRestore.cs`

karmeli87 commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs:
```diff
@@ -0,0 +1,120 @@
+
+        var context = JsBlittableBridge.Translate(Context, DocumentScript.ScriptEngine, args[0].AsObject());
+        string json = context.ToString();
+        string hash = AttachmentsStorageHelper.CalculateHash(MemoryMarshal.AsBytes(json.AsSpan()));
```
seems strange to find it under `AttachmentsStorageHelper`

karmeli87 commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,392 @@
+                        context.DocumentDatabase.DocumentsStorage.Put(context, document.Id, expectedChangeVector: null, document.Data);
+                    }
+
+                    foreach (var item in items)
+                    {
+                        hashes.Add(item.ContextOutput.AiHash);
+
+                        if (item?.ModelOutput is null)
+                            continue;
+
+                        var dvj = new DynamicJsonValue
+                        {
+                            ["output"] = item.ModelOutput.Output,
+                            ["input"] = item.ContextOutput.Context
+                        };
+
+                        var args = context.ReadObject(dvj, document.Id);
+                        var cmd = lastPatch = new PatchDocumentCommand(
+                            context: context,
+                            id: document.Id,
+                            expectedChangeVector: null,
+                            skipPatchIfChangeVectorMismatch: false,
+                            patch: (req, args),
+                            patchIfMissing: default,
+                            createIfMissing: null,
+                            identityPartsSeparator: Database.IdentityPartsSeparator,
+                            isTest: false,
+                            debugMode: false,
+                            collectResultsNeeded: true,
+                            returnDocument: false,
+                            ignoreMaxStepsForScript: false);
+
+                        cmd.Execute(context, recordingState: null);
+
+                        item.DebugActions = cmd.DebugActions;
+                        item.DebugOutput = cmd.DebugOutput;
+                    }
+
+                    if (lastPatch?.PatchResult?.ModifiedDocument != null)
+                        outputDocument = GenAiBatchPatchCommand.UpdateHashesInMetadata(document.Id, lastPatch.PatchResult.ModifiedDocument, Configuration.Name, hashes, context);
```
should be replaced by `GenAiBatchPatchCommand`?

karmeli87 commented on src/Raven.Server/ServerWide/ClusterStateMachine.License.cs:
```diff
@@ -174,8 +175,13 @@ private void AssertLicenseLimits(string type, ServerStore serverStore, DatabaseR
                 AssertDocumentsCompressionLicenseLimits(serverStore, databaseRecord, context);
                 break;
             case nameof(PutAiConnectionStringCommand):
+            case nameof(AddEmbeddingsGenerationCommand):
```
this looks strange... this not related to this feature? if so where was this before? @grisha-kotler you added that?

grisha-kotler replied:
Before this, if we don't have the Embeddings Generation feature, we would check the number of connection strings that aren't Embedded.
Since we use `PutAiConnectionStringCommand` also for Gen AI, we need to change the implementation.

karmeli87 commented on src/Raven.Server/ServerWide/ClusterStateMachine.License.cs:
```diff
@@ -932,17 +939,32 @@ private static void AssertEmbeddingsGeneration(DatabaseRecord databaseRecord, Li
+        if (databaseRecord.AiConnectionStrings.Count == 0 || databaseRecord.EmbeddingsGenerations.Count == 0)
+            return;
+
+        foreach (var config in databaseRecord.EmbeddingsGenerations)
```
same?

grisha-kotler replied:
https://github.com/ravendb/ravendb/pull/20580#discussion_r2094617399

kalczur committed:
RavenDB-23942 Fix virtual list after collapse

kalczur committed:
RavenDB-23942 Add hover on editor

M4xymm commented on src/Raven.Studio/typescript/components/pages/database/aiHub/aiTasks/AiTasks.tsx:
```diff
@@ -148,6 +158,8 @@ export default function AiTasks() {
         isDeleting,
     };
 
+    console.log("kalczur genAiTasks", genAiTasks);
```
`console.log` to remove

kalczur committed:
RavenDB-23942 Ignore story unhandled errors

M4xymm commented on src/Raven.Studio/typescript/components/pages/database/tasks/ongoingTasks/editTasks/editGenAiTask/partials/fields/EditGenAiTaskBasicFields.tsx:
```diff
@@ -0,0 +1,123 @@
+    const dispatch = useAppDispatch();
+
+    const isNewTask = useAppSelector(editGenAiTaskSelectors.isNewTask);
+    const isEncrypted = useAppSelector(databaseSelectors.activeDatabase)?.isEncrypted ?? false;
```
`?? false` isnt strictly necessary, undefined is already falsy
```suggestion
    const isEncrypted = useAppSelector(databaseSelectors.activeDatabase)?.isEncrypted;
```

M4xymm commented on src/Raven.Studio/typescript/components/pages/database/tasks/ongoingTasks/editTasks/editGenAiTask/partials/fields/EditGenAiTaskNodeField.tsx:
```diff
@@ -0,0 +1,76 @@
+
+    const nodes = useAppSelector(clusterSelectors.allNodes);
+
+    const possibleMentors = nodes.filter((x) => x.type === "Member").map((x) => x.nodeTag);
```
Imo, rename `possibleMentors` to `possibleMentorsOptions` and build mapped `[value,label]` array. It will work the same way, and it will be more readable.
```suggestion
    const possibleMentorsOptions = nodes
  .filter(node => node.type === 'Member')
  .map(node => ({ value: node.nodeTag, label: `Node ${node.nodeTag}` }));
```

M4xymm commented on src/Raven.Studio/typescript/components/pages/database/tasks/ongoingTasks/editTasks/editGenAiTask/partials/steps/EditGenAiTaskStepBasic.tsx:
```diff
@@ -0,0 +1,131 @@
+export function mapAiConnectionStringToSettingsDto(
+    connection: Raven.Client.Documents.Operations.AI.AiConnectionString
+): AiConnectionStringsSettings {
+    if (connection.AzureOpenAiSettings) {
+        return connection.AzureOpenAiSettings;
+    }
+    if (connection.GoogleSettings) {
+        return connection.GoogleSettings;
+    }
+    if (connection.HuggingFaceSettings) {
+        return connection.HuggingFaceSettings;
+    }
+    if (connection.OllamaSettings) {
+        return connection.OllamaSettings;
+    }
+    if (connection.EmbeddedSettings) {
+        return connection.EmbeddedSettings;
+    }
+    if (connection.OpenAiSettings) {
+        return connection.OpenAiSettings;
+    }
+    if (connection.MistralAiSettings) {
+        return connection.MistralAiSettings;
+    }
+
+    return null;
```
Perhaps it'd be cleaner to implement `mapAiConnectionStringToSettingsDto` like tihs:

```suggestion
return [
  connection.AzureOpenAiSettings,
  connection.GoogleSettings,
  connection.HuggingFaceSettings,
  connection.OllamaSettings,
  connection.EmbeddedSettings,
  connection.OpenAiSettings,
  connection.MistralAiSettings
].find(Boolean) ?? null;
```

M4xymm commented on src/Raven.Studio/typescript/components/pages/database/tasks/ongoingTasks/editTasks/editGenAiTask/partials/steps/EditGenAiTaskStepBasic.tsx:
```diff
@@ -0,0 +1,131 @@
+const getConnectorType = (
+    connection: Raven.Client.Documents.Operations.AI.AiConnectionString
+): Raven.Client.Documents.Operations.AI.AiConnectorType => {
+    if (connection.AzureOpenAiSettings) {
+        return "AzureOpenAi";
+    }
+    if (connection.GoogleSettings) {
+        return "Google";
+    }
+    if (connection.HuggingFaceSettings) {
+        return "HuggingFace";
+    }
+    if (connection.OllamaSettings) {
+        return "Ollama";
+    }
+    if (connection.EmbeddedSettings) {
+        return "Embedded";
+    }
+    if (connection.OpenAiSettings) {
+        return "OpenAi";
+    }
+    if (connection.MistralAiSettings) {
+        return "MistralAi";
+    }
+
+    return "None";
```
such a solution also as below will minimize the code and will be faster for possible expansion

```suggestion
    const settingsToTypeMap = {
        AzureOpenAiSettings: "AzureOpenAi",
        GoogleSettings: "Google",
        HuggingFaceSettings: "HuggingFace",
        OllamaSettings: "Ollama",
        EmbeddedSettings: "Embedded",
        OpenAiSettings: "OpenAi",
        MistralAiSettings: "MistralAi"
    } as const;
    
    const foundSetting = Object.keys(settingsToTypeMap).find(
        setting => connection[setting as keyof typeof connection]
    );
    
    return foundSetting 
        ? settingsToTypeMap[foundSetting as keyof typeof settingsToTypeMap] 
        : "None";
```

kalczur replied:
I don't like the `as keyof typeof` part. I will stick to simple if statement, it is easier to read. I will also throw an Error when setting is not handled

kalczur committed:
RavenDB-23942 Add schema provider button

kalczur committed:
RavenDB-23942 Styling

kalczur commented:
@maciejaszyk 
1. I can block Embedded connector for GenAI
2. Im using `/docs` EP with `startsWith`. Is there any option to get docs for a collection?

kalczur committed:
RavenDB-23942 Fix icon

ayende commented:
> 2. Im using `/docs` EP with `startsWith`. Is there any option to get docs for a collection?

What is it that you are trying to do?

kalczur commented:
> > 2. Im using `/docs` EP with `startsWith`. Is there any option to get docs for a collection?
> 
> What is it that you are trying to do?

It's related to @maciejaszyk comment: https://github.com/ravendb/ravendb/pull/20580#issuecomment-2871534402
He suggested showing only document IDs from a given collection

kalczur committed:
RavenDB-23942 Add 'Force sending cached data' toggle

kalczur committed:
RavenDB-23942 Fix mapping to dto

kalczur committed:
RavenDB-23942 Fix virtual list

Commit:
RavenDB-24096 : Gen-AI expose stats of the task

Lwiel commented on src/Raven.Server/Documents/AI/AiGen/ChatCompletionClient.cs:
```diff
@@ -0,0 +1,169 @@
+            {
+                new JsonObject
+                {
+                    ["role"] = "system",
```
Can we make constants for these? And do we always want to model it this way?

ayende replied:
Yes, that is the raw API that they support.

Lwiel commented on src/Raven.Server/Documents/AI/AiGen/ChatCompletionClient.cs:
```diff
@@ -0,0 +1,169 @@
+        using var stream = await reply.Content.ReadAsStreamAsync();
+        var response = await JsonDocument.ParseAsync(stream);
+
+        if (response.RootElement.TryGetProperty("choices", out var prop) is false)
```
Isn't this OpenAI specific schema?

ayende replied:
We are supporting _this_ API, nothing else. IT works with Ollama, Grok, etc.

Lwiel commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -260,6 +262,7 @@ private IEnumerable<EtlProcess> GetRelevantProcesses<T, TConnectionString>(List<
                 QueueEtlConfiguration queueConfig = null;
                 SnowflakeEtlConfiguration snowflakeConfig = null;
                 EmbeddingsGenerationConfiguration embeddingsGenerationConfig = null;
+                GenAiConfiguration aiGenGenerationConfig = null;
```
genAi

Lwiel commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -534,6 +549,7 @@ public void HandleDatabaseRecordChange(DatabaseRecord record)
             var myQueueEtl = new List<QueueEtlConfiguration>();
             var mySnowflakeEtl = new List<SnowflakeEtlConfiguration>();
             var myEmbeddingsGenerationEtl = new List<EmbeddingsGenerationConfiguration>();
+            var myGenAiEtl = new List<GenAiConfiguration>();
```
Not sure if we need `Etl` suffix here, same for embeddings generation

ppekrol replied:
myGenAis

Lwiel commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiResultItem.cs:
```diff
@@ -0,0 +1,58 @@
+        [nameof(Output)] = Output
+    };
+
+
```
Formatting

ppekrol replied:
ping

Lwiel commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,413 @@
+
+    protected override void EnterFallbackMode(Exception e, DateTime? lastErrorTime)
+    {
+        // rate limits in embeddings are usually expressed as requests per minute or tokens per minute
```
in embeddings -> for chat completions

kalczur committed:
RavenDB-23942 Improve code quality

kalczur committed:
RavenDB-23942 Fix unhandled ace mode

kalczur committed:
RavenDB-23942 Add collapse to summary

Lwiel commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/Stats/GenAiOperations.cs:
```diff
@@ -0,0 +1,6 @@
+
+public sealed class GenAiOperations
+{
+    public const string LoadToModel = "Model/Load";
```
Do we want to expose time spent on waiting because of rate limits?

ayende replied:
Yes, but that isn't showing here. It is shown in the time _between_ batches

kalczur committed:
RavenDB-23942 Handle non existing task id

kalczur committed:
RavenDB-23942 Update default values

kalczur committed:
RavenDB-23942 Add GenAI link to omni search

Lwiel commented on src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:
```diff
@@ -0,0 +1,413 @@
+        {
+            AiConnectorType.Ollama => (cfg.Connection.OllamaSettings.Uri, cfg.Connection.OllamaSettings.Model, (string)null),
+            AiConnectorType.OpenAi => (cfg.Connection.OpenAiSettings.Endpoint, cfg.Connection.OpenAiSettings.Model, cfg.Connection.OpenAiSettings.ApiKey),
+            _ => throw new NotSupportedException(connectorType.ToString())
```
Can we have more descriptive message here?

ayende replied:
We need an exchaustive impl here.

Lwiel commented on src/Raven.Server/Documents/Sharding/Handlers/ShardedGenAiHandler.cs:
```diff
@@ -0,0 +1,15 @@
+
+public sealed class ShardedGenAiHandler : ShardedDatabaseRequestHandler
+{
+    [RavenShardedAction("/databases/*/admin/ai/genai/test", "POST")]
```
`gen-ai`?

Lwiel commented on src/Raven.Server/ServerWide/ClusterCommandsVersionManager.cs:
```diff
@@ -202,6 +202,10 @@ public sealed class ClusterCommandsVersionManager
+            
+            [nameof(AddGenAiCommand)] = 70_002,
+            [nameof(UpdateGenAiCommand)] = 70_002,
+
```
Do we support removing these tasks?

ayende replied:
Yes, but that one is generic, no?

shaharhikri committed:
RavenDB-24123 - GenAI: Create a client for LLM API

kalczur committed:
RavenDB-23942 Allow to add and delete context and model output

kalczur committed:
RavenDB-23942 Add identifier

kalczur committed:
RavenDB-23942 Add sample object validation

ppekrol commented on src/Raven.Client/Documents/Operations/AI/GenAiConfiguration.cs:
```diff
@@ -0,0 +1,110 @@
+    public override bool Validate(out List<string> errors, bool validateName = true, bool validateConnection = true)
+    {
+        if (validateConnection && Initialized == false)
+            throw new InvalidOperationException("AiGen configuration must be initialized");
```
Gen AI

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+        }
+
+        // As double: 33011.382867097855
+        if (double.TryParse(input, out var seconds2))
```
just to be on the safe side, I would add InvariantCulture

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/GenAiExceptions.cs:
```diff
@@ -0,0 +1,56 @@
+    {
+    }
+
+    public class GenUnsuccessfulRequestException : GenAiException
```
GenAi

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/GenAiExceptions.cs:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    /// <summary>HTTP 429 – too many requests.</summary>
+    public sealed class GenAiRateLimitException(string message) : GenAiTooManyRequestsException(message)
```
should it have GenAi prefix? It will be used as well in Embeddings Generation, it is just a general exception from the client

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+    internal static string GetAllowedUniqueName(string schemaOrSampleObject)
+    {
+        var originalHash = AttachmentsStorageHelper.CalculateHash(MemoryMarshal.AsBytes(schemaOrSampleObject.AsSpan()));
+        byte[] hashBytes = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(originalHash));
```
you need to dispose the SHA256 IIRC

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+
+        return JsonSerializer.Serialize(schema, new JsonSerializerOptions { WriteIndented = true });
+
+        JsonObject GenerateJsonObjectFromSampleObject(JsonElement element)
```
JsonSchemaObject

shaharhikri replied:
Done

ayende commented:
@Lwiel  about https://github.com/ravendb/ravendb/pull/20580#discussion_r2086769830
What do you mean OpenAI has two APIs?

There is the chat completion API, and nothing else that I'm aware of.
Unless you are talking about the Assistants API? That is similar, but quite different.

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+using JsonSerializer = System.Text.Json.JsonSerializer;
+
+
+namespace Raven.Server.Documents.AI.AiGen;
```
shouldn't this be moved one level up, dont we want at the end to also add the ability to generate embeddings and at the end substitute the semantic kernel?

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+    public AbstractChatCompletionClient(Uri baseUri, string model, string apiKey, string structuredOutputSchema)
+    {
+        _model = model;
+        _client = new()
```
If we would take leverage from RavenServerHttpClientFactory (need to expose probably a new method there), then we you would not have to deal with handling of DNS and other transient errors

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+        }, Conventions);
+
+        
+        content.Headers.Add("Content-Type", "application/json");
```
use constant, also maybe better to set default header?

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+        _client = new()
+        {
+            BaseAddress = baseUri,
+            DefaultRequestHeaders =
```
should we also consider sending compressed content? is there a support for it?

shaharhikri replied:
Done, adding the ability to pass DocumentConventions (with UseHttpCompression true).
It false by default because openai doesn't work with compression..

ppekrol commented on src/Raven.Server/Documents/AI/AiGen/AbstractChatCompletionClient.cs:
```diff
@@ -0,0 +1,458 @@
+    private readonly string _model;
+    private readonly HttpClient _client;
+    private readonly string _structuredOutputSchema;
+    private readonly JsonContextPool _contextPool = new JsonContextPool();
```
shouldnt' we inject that and use e.g. context pool from the database?

shaharhikri replied:
Done

Commit:
RavenDB-24189 : base the hash also on the configuration (prompt + schema + update), not just the context object

kalczur committed:
RavenDB-23942 Add load from file

kalczur committed:
RavenDB-23942 Fix naming and test button disabled conditions

shaharhikri committed:
RavenDB-24123 - Address PR Comments, Stop using SHA-256 in GetAllowedUniqueName

grisha-kotler committed:
RavenDB-23942 - enable the license attribute

kalczur committed:
RavenDB-23942 Add AceUnifiedDiff

kalczur committed:
RavenDB-23942 Use AceUnifiedDiff in test results

kalczur committed:
RavenDB-23942 Fix playground tab navigation

kalczur committed:
RavenDB-23942 Styling

kalczur committed:
RavenDB-23942 Add info hub and remove HStack and VStack

kalczur committed:
RavenDB-23942 Prepare ace editor

kalczur committed:
RavenDB-23942 Move ace to separate folder

kalczur committed:
RavenDB-23942 Provide ace editor context

kalczur committed:
RavenDB-23942 Styling ace actions

kalczur committed:
RavenDB-23942 Change label text

kalczur committed:
RavenDB-23942 Add ace format action

Commit:
RavenDB-23942 : bug fix - don't try to parse SampleObject into JsonSchema when we're in initial test mode (sample object/schema is not guaranteed at this point, it is provided in later test stages)

kalczur committed:
RavenDB-23942 Add title

kalczur committed:
RavenDB-23942 Add upload and delete action

kalczur committed:
RavenDB-23942 Text fixes

kalczur committed:
RavenDB-23942 Omit aceRef in AceUnifiedDiff

kalczur committed:
RavenDB-23942 Move edit warning to tooltip

kalczur committed:
RavenDB-23942 Add syntax help

kalczur committed:
RavenDB-23942 Icon position

shaharhikri committed:
RavenDB-24123 - Make regex compile on the compilation time, instead of run time

Commit:
RavenDB-24171 : GenAI error handling

Commit:
RavenDB-24171 : add error-handling tests

kalczur committed:
RavenDB-23942 Styling

kalczur committed:
RavenDB-23942 Add EnterpriseAi to About page

kalczur committed:
RavenDB-23942 Prepare availability summary

kalczur committed:
RavenDB-23942 Summary styling

shaharhikri committed:
RavenDB-24188 - GenAi tests attribute

Danielle9897 committed:
RavenDB-24256 "About this view" text

Danielle9897 committed:
RavenDB-24256 "About this view" text - small fix

Commit:
RavenDB-24094 : Gen AI: Integrate to backup/restore

Commit:
RavenDB-24094 : add backup/restore tests

Commit:
RavenDB-23942 : fix failing tests

shaharhikri committed:
RavenDB-24188 - Use RavenGenAiData attribute on GenAi tests

kalczur committed:
RavenDB-23942 Summary styling

kalczur committed:
RavenDB-23942 Widget styling

Commit:
RavenDB-23942 : add genAiCount to GetOngoingTasksInfoItem

kalczur committed:
RavenDB-23942 Fix bug with override

kalczur committed:
RavenDB-23942 Add badge to add new task view

kalczur committed:
RavenDB-23942 Remove old tests

kalczur committed:
RavenDB-23942 Add license restrictions for edit GenAI view

kalczur committed:
RavenDB-23942 Fix lint and tests

kalczur committed:
RavenDB-23942 Open License comparison by default when no feature in license

mateuszbartosik committed:
RavenDB-23942 Fix colors for Ongoing tasks & icon for Licensing modal

kalczur committed:
RavenDB-23942 Fix popover

kalczur committed:
RavenDB-23942 Fix prettier

Commit:
RavenDB-24173 : address some PR comments

kalczur committed:
RavenDB-23942 Mock license

Commit:
RavenDB-24173 : fix test files after rebase

kalczur committed:
RavenDB-23942 Fix autocomplete on select

Commit:
RavenDB-24171 Error handling
- Restructured how we are writing to the database, only write once per doc (even if shows up multiple times)
- Simplified error handling for update script failures

Commit:
RavenDB-24189
- Will take into account, Prompt, UpdateScript, JsonSchema and connection string name as part of the context hash
- Partially cache the hash computation to improve performance

Commit:
RavenDB-24172 : limit the number of concurrent requests
- Adding max concurrency handling to limit model load
- If there is an error, will reset concurrency limit to 1 and slowly raise it higher to the limit

Commit:
RavenDB-24173 : bug fixes
- fix NRE in UpdateHashString (in initial phase of test-mode the prompt/update/schema can be empty)
- avoid an infinite loop - don't update metadata hashes if everything is the same

Commit:
RavenDB-24173 : skip flaky tests

Commit:
RavenDB-24173: Fix incorrect item-task alignment in SendToModel

The items list passed to ProcessModelResults must match the tasks list by index, since we rely on this ordering to assign model results to the correct GenAiResultItem.

Previously, we skipped cached items but still added tasks for non-cached items without updating the items list accordingly. This caused a mismatch: tasks[x] no longer corresponded to items[x], leading to incorrect result assignments and potential data corruption.

This change ensures that we preserve a strict one-to-one mapping between items and tasks.

Commit:
RavenDB-24280 : rename 'context()' to 'ai.genContext()'

ppekrol committed:
RavenDB-23942 removed ChatCompletion from SemanticKernel

ppekrol committed:
RavenDB-23942 more fixes

kalczur committed:
RavenDB-23942 Fix studio build

kalczur committed:
RavenDB-23942 Generate identifier on blur

kalczur committed:
RavenDB-23942 Wrap prompt

kalczur committed:
RavenDB-23942 Use select autocomplete

kalczur committed:
RavenDB-23942 Update text

Commit:
RavenDB-24173 : bug fix in JsBlittableBridge.Translate - need to clear cached-properties before writing new document

kalczur committed:
RavenDB-23942 Fix update script test

kalczur committed:
RavenDB-23942 Update sample scripts in storybook

kalczur committed:
RavenDB-23942 Fix tooltip text

kalczur committed:
RavenDB-23942 Fix task list info error

kalczur committed:
RavenDB-23942 Fix dev license override

kalczur committed:
RavenDB-23942 Add dirty flag

kalczur committed:
RavenDB-23942 Add llm icon

Commit:
RavenDB-24235 : GenAI Test SendToModel shouldnt require a Document

Commit:
RavenDB-24173 : fix failing tests

Commit:
RavenDB-24173 : skip flaky test

kalczur committed:
RavenDB-23942 Filter connectors for GenAI task

kalczur committed:
RavenDB-23942 Send empty document on SendToModel stage

ppekrol commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1278,6 +1283,12 @@ private void ThrowIfCannotActivateLicense(LicenseStatus newLicenseStatus)
 
+            if (genAiCount > 0 && newLicenseStatus.HasGenAi == false)
+            {
+                var message = GenerateDetails(genAiCount, "AI Generation");
```
feature name is GenAI

ppekrol commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1651,13 +1662,25 @@ public void AssertCanAddEmbeddingsGenerationTask(AiConnectionString aiConnection
+            if (LicenseStatus.HasGenAi)
+                return;
+
+            const string message = "Your current license doesn't include the Gen AI feature";
```
GenAI

kalczur committed:
RavenDB-23942 Gen AI -> GenAI

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.License.cs:
```diff
@@ -932,17 +939,32 @@ private static void AssertEmbeddingsGeneration(DatabaseRecord databaseRecord, Li
-        
-        throw new LicenseLimitException(LimitType.EmbeddingsGeneration, "Your license doesn't support using the Embeddings Generation feature.");
+
+        throw new LicenseLimitException(LimitType.GenAi, "Your license doesn't support using the AI Generation feature.");
```
GenAI

ppekrol commented on src/Raven.Server/ServerWide/Commands/AI/AddGenAiCommand.cs:
```diff
@@ -0,0 +1,56 @@
+    {
+        return;
+
+        // TODO
```
TODO

ppekrol commented on src/Raven.Studio/typescript/components/common/LicenseRestrictedBadge.tsx:
```diff
@@ -4,7 +4,7 @@ import classNames from "classnames";
 import { licenseSelectors } from "./shell/licenseSlice";
 
-export type LicenseBadgeText = "Professional +" | "Enterprise";
+export type LicenseBadgeText = "Professional +" | "Enterprise" | "Enterprise Ai";
```
AI @kalczur

ppekrol commented on src/Raven.Studio/typescript/components/common/LicenseRestrictedBadge.tsx:
```diff
@@ -34,6 +36,8 @@ function getClassName(licenseBadgeText: LicenseBadgeText, isCloud: boolean): "en
             return "enterprise";
         case "Professional +":
             return "professional";
+        case "Enterprise Ai":
```
AI @kalczur

Commit:
RavenDB-24173 : rename GenAiEtls -> GenAis

kalczur committed:
RavenDB-23942 Enterprise Ai -> Enterprise AI

karmeli87 committed:
RavenDB-23942 limit the extraction of genai to `Configuration.MaxConcurrency * 8`

kalczur committed:
RavenDB-24296 Show GenAI in stats

kalczur committed:
RavenDB-24304 Add GenAI toggle to import/export
